File: js/src/CustomDS/RandomizedSet.js


export class RandomizedSet{

	#map;
	#list;

	constructor(){
		this.#map = {};
		this.#list = [];
	}


	insert(value){

		if(this.#map.hasOwnProperty(value)) return false;

		this.#map[value] = this.#list.length;

		this.#list.push(value);

		return true;
	}

	remove(value){

		if (!this.#map.hasOwnProperty(value)) return false;

		let dIndex = this.#map[value];
		
		let lastEle = this.#list[this.#list.length - 1];

		if(dIndex !== this.#list.length -1){

			this.#map[value] = dIndex;

			this.#list[dIndex] = lastEle;

		}

		this.#list.pop();

		delete this.#map[value];

		return true;

	}

	Random(){

		let rand = Math.floor(Math.random() * this.#list.length);
		return this.#list[rand];
	}
}

File: python/custom_ds/randomized_set.py
import random

class RandomizeSet():

	def __init__(self):
		
		self.list = []
		self.map = {}



	def insert(self, value):

		if value in self.map:

			return False

		self.map[value] = len(self.map);

		self.list.append(value);

		return True


	def remove(self,value):

		if value not in self.list:

			return False

		dIndex = self.map[value]
		lastEle = self.list[-1]

		if dIndex != len(self.list) -1:
			self.map[lastEle] = dIndex
			self.list[dIndex] = lastEle

		del self.map[value]
		self.list.pop()

		return True


	def random(self):

		return random.choice(self.list);
	

		